### PR TITLE
[Runtime] Refactor _gatherWrittenGenericParameters

### PR DESCRIPTION
**Pull request by Azoy**
There are 2 of the same implementation of `_gatherWrittenGenericParameters` in MetadataLookup as well as `_buildDemanglingForGenericArgs` in Demangle. Unify these implementations. This also fixes a slight bug in the following:

```swift
struct Generic<T, U, V> {}

extension Generic where T == U, U == V.Element, V: Collection {
  struct Nested {}
}

print(_typeName(Generic<Int, Int, [Int]>.Nested.self)) // << invalid type >>
```

where we assumed the rhs parameter is a generic argument with (depth, index) and fail if it isn't. This is incorrect because the rhs can totally be one of these same type constraints to associated types.

### CODE REVIEW

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please test

File: stdlib/public/runtime/MetadataLookup.cpp [hunk 1]
```diff
@@ -1278,10 +1273,41 @@ _gatherGenericParameters(const ContextDescriptor *context,
...
     llvm::SmallVector<MetadataOrPack, 8> allGenericArgs;
 
-    // If we have a parent, gather it's generic arguments "as written".
-    if (parent) {
-      _gatherWrittenGenericArgs(parent, parent->getTypeContextDescriptor(),
-                                allGenericArgs, demangler);
+    auto generics = context->getGenericContext();
+    assert(generics);
+
+    // If we have a parent, gather it's generic arguments "as written". If our
```
Comment (glessard) [inline stdlib/public/runtime/MetadataLookup.cpp:1279]:
```suggestion
    // If we have a parent, gather its generic arguments "as written". If our
```

File: stdlib/public/runtime/MetadataLookup.cpp [hunk 2]
```diff
@@ -3232,76 +3265,72 @@ demangleToGenericParamRef(StringRef typeName) {
...
+      genericArgs.push_back(MetadataOrPack(keyArgs[argIndex]));
+
+      argIndex += 1;
+    } else {
+      // Leave a gap for us to fill in by looking at same-type requirements.
+      genericArgs.push_back(MetadataOrPack());
+      missingWrittenArguments = true;
+    }
+
     switch (param.getKind()) {
```
Comment (slavapestov) [inline stdlib/public/runtime/MetadataLookup.cpp:3300]: Maybe just remove this switch statement or replace with an assert

File: stdlib/public/runtime/MetadataLookup.cpp [hunk 3]
```diff
@@ -3318,44 +3347,106 @@ static void _gatherWrittenGenericArgs(
...
+
+    // If we already have an argument for the lhs, then just check that it is
+    // indeed == to the rhs type.
+    if (auto genericArg = genericArgs[*lhsFlatIndex]) {
+      if (genericArg.getMetadata() != rhsParam) {
+        return false;
+      }
+    } else {
+      // If we don't have a lhs yet, then it's just the rhs.
+      genericArgs[*lhsFlatIndex] = MetadataOrPack(rhsParam);
```
Comment (slavapestov) [inline stdlib/public/runtime/MetadataLookup.cpp:3444]: How can this happen?

File: stdlib/public/runtime/MetadataLookup.cpp [hunk 4]
```diff
@@ -3318,44 +3347,106 @@ static void _gatherWrittenGenericArgs(
...
+
+    if (!rhsParam) {
+      return false;
+    }
+
+    // If we already have an argument for the lhs, then just check that it is
+    // indeed == to the rhs type.
+    if (auto genericArg = genericArgs[*lhsFlatIndex]) {
+      if (genericArg.getMetadata() != rhsParam) {
+        return false;
```
Comment (slavapestov) [inline stdlib/public/runtime/MetadataLookup.cpp:3440]: Aren't callers supposed to check the requirements with checkGenericRequirements()?

File: stdlib/public/runtime/MetadataLookup.cpp [hunk 5]
```diff
@@ -3318,44 +3347,106 @@ static void _gatherWrittenGenericArgs(
...
+
+    genericArgs[*rhsFlatIndex] = genericArgs[*lhsFlatIndex];
+  }
+
+  if (!needToCheckAssociatedSameTypes) {
+    return true;
+  }
+
+  // Once again, loop through our list and look for same type constraints where
+  // the rhs is an associated type of sorts.
```
Comment (slavapestov) [inline stdlib/public/runtime/MetadataLookup.cpp:3402]: I believe it should be possible to populate all genericArgs in one pass. A generic parameter is non-canonical if you have a requirement `T == ConcreteType` or `U == T` (note the order) where `U` is another generic parameter.

`T.A.B == U` does not make `U` non-canonical on the other hand, so it should not be necessary to populate `U` here.

**Comment by slavapestov** [comment]
> extension Generic where T == U, U == V.Element, V: Collection {

In this example, T and V are canonical, while U is not. However, U is made non-canonical by `T == U` and not `U == V.Element`. So if we instead had

> extension Generic where U == V.Element, V: Collection {

Then now, all three of `T`, `U` and `V` would be canonical.

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@slavapestov Ok, yeah we didn't need the second loop at all. If the rhs is an associated type then we can just skip the requirement entirely because the lhs has already been filled in and that's all we care about.

**Review by slavapestov — APPROVED** [review]
Great work!

**Pull request merged**